src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1109,23 +1144,59 @@ void CASESession::SendErrorMsg(SigmaErrorType errorCode)
                    ChipLogError(SecureChannel, "Failed to send error message"));
 }
 
-CHIP_ERROR CASESession::FindValidTrustedRoot(const System::PacketBufferTLVReader & tlvReader, uint32_t nTrustedRoots)
+CHIP_ERROR CASESession::GenerateDestinationID(const ByteSpan & random, const Credentials::CertificateKeyId * trustedRootId,
+                                              NodeId nodeId, MutableByteSpan & destinationId)
 {
-    CertificateKeyId trustedRoot;
-    System::PacketBufferTLVReader suppTlvReader;
-    uint8_t trustedRootId[kTrustedRootIdSize];
+    HKDF_sha_crypto mHKDF;
+    // TODO: read fabricId
+    uint64_t fabricId              = 0;
+    uint16_t destinationMessageLen = kSigmaParamRandomNumberSize + kP256_PublicKey_Length + sizeof(fabricId) + sizeof(nodeId);
+    System::PacketBufferHandle destinationMessage = System::PacketBufferHandle::New(destinationMessageLen);
+    const ChipCertificateData * rootCertificate   = mOpCredSet->GetRootCertificate(*trustedRootId);
+    VerifyOrReturnError(!destinationMessage.IsNull(), CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(rootCertificate != nullptr, CHIP_ERROR_CERT_NOT_FOUND);
+    VerifyOrReturnError(!rootCertificate->mPublicKey.empty(), CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(rootCertificate->mPublicKey.size() == kP256_PublicKey_Length, CHIP_ERROR_INTERNAL);
+
+    Encoding::LittleEndian::BufferWriter bbuf(destinationMessage->Start(), destinationMessageLen);
+
+    bbuf.Put(random.data(), random.size());
+    bbuf.Put(rootCertificate->mPublicKey.data(), rootCertificate->mPublicKey.size());
+    bbuf.Put64(fabricId);
+    bbuf.Put64(nodeId);
+
+    VerifyOrReturnError(bbuf.Fit(), CHIP_ERROR_NO_MEMORY);
+
+    ReturnErrorOnFailure(mHKDF.HKDF_SHA256(destinationMessage->Start(), destinationMessageLen, 0, 0, mIPK, sizeof(mIPK),
```

Comment:
There is one choice of IPK for outgoing destination ID (the current one chosen by initator), but this method will eventually need to be called several times, one for each IPK in force (where there could be rotation at play) at the receiver.

Strongly suggest making this "compute-only" method stateless, by:

1- Adding a FabricID argument
2- Adding an IPK argument

Then the `GenerateDestinationID` could also be validated by unit test against the canonical test vector in Matter spec.

Reply:
The function used here is an HKDF, rather than the specified HMAC. They are very closely related, but not the same.

The algorithm used here mixes and matches fields in a way that does not match spec. See https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/crypto_primitives/case_destination_id_test_vector.py for pseudo-code of a destination ID.

Your version should yield the *exact same* output for the same inputs. It will not.

I implemented your version quickly in the test vector (see attached) and given the same inputs, it does not match the output of HMAC-SHA256 due to how the data is preprocessed for HKDF purposes.

This means that the destination ID computation is not to-spec.

Suggested fix:

1- Create a unit test case that matches the destination ID test vector from spec
2- Add CHIP_Crypto_HMAC to CryptoPAL (not yet implemented, but easy to implement and verify). The semantics are expressed in chapter 3 of the spec
3- Update the logic here to use CHIP_Crypto_HMAC.

I can assist, as I wrote the test vectors in the spec and designed the destinationId method.

Reply:
I started the implementation of CHIP_Crypto_HMAC in CHIPCryptoPAL

Reply:
Updated to use HMAC.